controllers/drain_controller.go

```diff
@@ -0,0 +1,492 @@
+// move the current state of the cluster closer to the desired state.
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (dr *DrainReconcile) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// try to lock the node to this reconcile loop.
+	// if we are not able this means there is another loop already handling this node, so we just exist
+	if !dr.TryLockNode(req.Name) {
+		return ctrl.Result{}, nil
+	}
+
+	// configure logs
+	reqLogger := log.FromContext(ctx)
+	reqLogger.Info("Reconciling Drain")
+	reqLogger.V(2).Info("node locked for drain controller", "nodeName", req.Name)
+
+	// we send to another function so the operator can have a defer function to release the lock
+	return dr.reconcile(ctx, req)
+}
+
+func (dr *DrainReconcile) reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// remove the lock when we exist the function
+	defer dr.unlockNode(req.Name)
+
+	req.Namespace = vars.Namespace
+
+	// configure logs
+	reqLogger := log.FromContext(ctx)
+
+	// get node object
+	node := &corev1.Node{}
+	err := dr.Get(ctx, req.NamespacedName, node)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			reqLogger.Error(err, "node object doesn't exist", "nodeName", req.Name)
+			return ctrl.Result{}, nil
+		}
+
+		reqLogger.Error(err, "failed to get node object from api re-queue the request", "nodeName", req.Name)
+		return ctrl.Result{}, err
+	}
+
+	// get sriovNodeNodeState object
+	nodeNetworkState := &sriovnetworkv1.SriovNetworkNodeState{}
+	err = dr.Get(ctx, req.NamespacedName, nodeNetworkState)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			reqLogger.Error(err, "sriovNetworkNodeState object doesn't exist", "nodeName", req.Name)
+			return ctrl.Result{}, nil
+		}
+
+		reqLogger.Error(err, "failed to get sriovNetworkNodeState object from api re-queue the request", "nodeName", req.Name)
+		return ctrl.Result{}, err
+	}
+
+	// create the drain state annotation if it doesn't exist in the sriovNetworkNodeState object
+	nodeStateDrainAnnotationCurrent, NodeStateDrainAnnotationCurrentExist := nodeNetworkState.Annotations[constants.NodeStateDrainAnnotationCurrent]
+	if !NodeStateDrainAnnotationCurrentExist {
+		err = utils.AnnotateObject(nodeNetworkState, constants.NodeStateDrainAnnotationCurrent, constants.DrainIdle, dr.Client)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		nodeStateDrainAnnotationCurrent = constants.DrainIdle
+	}
+
+	// create the drain state annotation if it doesn't exist in the node object
+	nodeDrainAnnotation, nodeDrainAnnotationExist := node.Annotations[constants.NodeDrainAnnotation]
+	if !nodeDrainAnnotationExist {
+		err = utils.AnnotateObject(node, constants.NodeDrainAnnotation, constants.DrainIdle, dr.Client)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		nodeDrainAnnotation = constants.DrainIdle
+	}
```

Comment:
Why do we have the annotation twice? Once on the node and once on the node state?

Reply:
that is right. one is the req from the daemon and the other one is the state from the operator point of view